packages/gatsby/src/utils/webpack/force-css-hmr-for-edge-cases.ts

```diff
@@ -0,0 +1,97 @@
+import { Compiler, Module } from "webpack"
+
+/**
+ * This is total hack that is meant to handle:
+ *  - https://github.com/webpack-contrib/mini-css-extract-plugin/issues/706
+ *  - https://github.com/webpack-contrib/mini-css-extract-plugin/issues/708
+ * The way it works it is looking up what HotModuleReplacementPlugin checks internally
+ * and tricks it by checking up if any modules that uses mini-css-extract-plugin
+ * changed or was newly added and then modifying blank.css hash.
+ * blank.css is css module that is used by all pages and is there from the start
+ * so changing hash of that _should_ ensure that:
+ *  - when new css is imported it will reload css
+ *  - when css imported by not loaded (by runtime) page template changes it will reload css
+ */
+export class ForceCssHMRForEdgeCases {
+  private name: string
+  private originalBlankCssHash: string
+  private blankCssKey: string
+  private hackCounter = 0
+  private previouslySeenCss: Set<string> = new Set<string>()
+
+  constructor() {
+    this.name = `ForceCssHMRForEdgeCases`
+  }
+
+  apply(compiler: Compiler): void {
+    compiler.hooks.thisCompilation.tap(this.name, compilation => {
+      compilation.hooks.fullHash.tap(this.name, () => {
```

Comment:
this entire hack is largely based on checking what HMR internal webpack plugin does and "tricking it" into thinking that `blank.css` file we import () is changed when:
 a) there is new css module or hash of already existing one changed (we set `anyCssChanged` to `true`)
 b) when css module that was in previous compilation is no longer there (that's combo of `seenCssInThisCompilation` and `this.previouslySeenCss`)